Fix Kodi specific services registry and add descriptions

azogue opened the pull request:
## Description:
 - Fixes issue #7528 when registering services for multiple Kodi media players.
 - Add descriptions for Kodi specific services: `media_player.kodi_set_shuffle` and `media_player.kodi_add_to_playlist`.
 - More error handling in Kodi API errors.
 - Makes compatible the existent specific service `media_player.kodi_set_shuffle` with the general `media_player.shuffle_set` service (both use the same method but with different named parameter, I think the Kodi specific service should be eliminated, since it is not)
 - Add `SUPPORT_SHUFFLE_SET`media player flag.

**Related issue (if applicable):** fixes #7528

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2614

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

azogue committed:
Fix Kodi specific services, add descriptions, add more handled exceptions

 - Fixes issue #7528
 - Add descriptions for Kodi specific services in services.yaml.
 - Error handling in Kodi API errors.
 - Make compatible the existent specific service `media_player.kodi_set_shuffle` with the general `media_player.shuffle_set` service (both use the same method but with different named parameter, I think the Kodi specific service should be eliminated, since it is not)

azogue committed:
fix line too long

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -78,22 +81,25 @@
 
 SERVICE_ADD_MEDIA = 'kodi_add_to_playlist'
-SERVICE_SET_SHUFFLE = 'kodi_set_shuffle'
+SERVICE_SET_SHUFFLE = 'kodi_set_shuffle'  # this is the same as mp.shuffle_set
```
Let's just eliminate this. It only existed for a short while before core shuffle support was added.

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -109,6 +115,8 @@
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the Kodi platform."""
+    if DATA_KODI not in hass.data:
+        hass.data[DATA_KODI] = []
```
Let's instead make this a dict with the entity_id as the key, so we don't have to search it later.

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -130,7 +138,11 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    descriptions = yield from hass.loop.run_in_executor(
+        None, load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
 
```
I think we want to add a guard clause here to check if the services are already registered. Something like
```python
if hass.services.has_service(DOMAIN, SERVICE_ADD_MEDIA):
    return
```

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -141,23 +153,30 @@ def async_service_handler(service):
     for service in SERVICE_TO_METHOD:
-        schema = SERVICE_TO_METHOD[service].get(
-            'schema', MEDIA_PLAYER_SCHEMA)
+        schema = SERVICE_TO_METHOD[service].get('schema')
```
You could even make this `schema = SERVICE_TO_METHOD[service]['schema']`. We want to ensure the schema is defined for future changes.

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -78,22 +81,25 @@
     vol.Optional(ATTR_MEDIA_ID): cv.string,
-    vol.Optional(ATTR_MEDIA_NAME): cv.string,
-    vol.Optional(ATTR_MEDIA_ARTIST_NAME): cv.string,
+    vol.Inclusive(ATTR_MEDIA_NAME, 'media_search'): cv.string,
```
Can we keep these optional, and default it to `ALL` in the service if looking for artists? Each media type should be able to make the best decision what to do if one isn't defined.

emlove reviewed (commented):
Looks good. Added a few notes.

azogue committed:
removed new services (for another PR); removed `kodi_set_shuffle` service

armills commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -130,6 +130,8 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
 
+    key = 'media_player.' + slugify(config.get(CONF_NAME))
+    hass.data[DATA_KODI][key] = entity
     async_add_devices([entity], update_before_add=True)
```
Somehow this was missed before, but this is a coroutine. If we make this `yield from async_add_devices(...`, the `entity_id` should be available afterwards. Use `entity.entity_id` as the dict key.

azogue replied:
I think it is not possible, since v0.40, the passed in `async_add_devices` method is a callback instead of a coroutine function. (https://home-assistant.io/blog/2017/03/11/repurpose-any-android-phone-as-ip-camera/)

I also did not like having to use `slugify` in that place, maybe it would be better to leave `hass.data[DATA_KODI]` as a list?

armills replied:
Ah, I missed that. In that case then yeah, going back to the list seems to make the most sense.

azogue committed:
requested changes

 - Removed `kodi_set_shuffle` service.
 - Optional `media_name` and `artist_name` parameters. `media_name` defaults to 'ALL'.
 - Guard clause to check if the services are already registered.

emlove reviewed (approved):
LGTM
